php-checks/src/main/java/org/sonar/php/checks/UseOfUninitializedVariableCheck.java

```diff
@@ -0,0 +1,279 @@
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.php.checks.utils.CheckUtils;
+import org.sonar.plugins.php.api.tree.Tree;
+import org.sonar.plugins.php.api.tree.Tree.Kind;
+import org.sonar.plugins.php.api.tree.declaration.ClassDeclarationTree;
+import org.sonar.plugins.php.api.tree.declaration.FunctionDeclarationTree;
+import org.sonar.plugins.php.api.tree.declaration.FunctionTree;
+import org.sonar.plugins.php.api.tree.declaration.MethodDeclarationTree;
+import org.sonar.plugins.php.api.tree.expression.AnonymousClassTree;
+import org.sonar.plugins.php.api.tree.expression.ArrayAccessTree;
+import org.sonar.plugins.php.api.tree.expression.AssignmentExpressionTree;
+import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
+import org.sonar.plugins.php.api.tree.expression.FunctionExpressionTree;
+import org.sonar.plugins.php.api.tree.expression.LexicalVariablesTree;
+import org.sonar.plugins.php.api.tree.expression.MemberAccessTree;
+import org.sonar.plugins.php.api.tree.expression.VariableIdentifierTree;
+import org.sonar.plugins.php.api.tree.expression.VariableTree;
+import org.sonar.plugins.php.api.tree.statement.ForEachStatementTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+
+@Rule(key = "S836")
+public class UseOfUninitializedVariableCheck extends PHPVisitorCheck {
+
+  private static final Set<Kind> PARENT_INITIALIZATION_KIND = EnumSet.of(
+    // Note: LEXICAL_VARIABLES are both, read and write, see VariableVisitor#visitFunctionExpression
+    Kind.PARAMETER,
+    Kind.GLOBAL_STATEMENT,
+    Kind.VARIABLE_DECLARATION,
+    Kind.EXPRESSION_STATEMENT,
+    Kind.REFERENCE_VARIABLE,
+    Kind.ARRAY_ASSIGNMENT_PATTERN_ELEMENT,
+    Kind.UNSET_VARIABLE_STATEMENT,
+    // CatchBlockTree#variable
+    Kind.CATCH_BLOCK,
+    Kind.ASSIGNMENT_BY_REFERENCE);
+
+  private static final Set<Kind> ASSIGNMENT_KIND = EnumSet.of(
+    Kind.AND_ASSIGNMENT,
+    Kind.ASSIGNMENT,
+    Kind.CONCATENATION_ASSIGNMENT,
+    Kind.DIVIDE_ASSIGNMENT,
+    Kind.LEFT_SHIFT_ASSIGNMENT,
+    Kind.MINUS_ASSIGNMENT,
+    Kind.MULTIPLY_ASSIGNMENT,
+    Kind.OR_ASSIGNMENT,
+    Kind.PLUS_ASSIGNMENT,
+    Kind.POWER_ASSIGNMENT,
+    Kind.REMAINDER_ASSIGNMENT,
+    Kind.RIGHT_SHIFT_ASSIGNMENT,
+    Kind.XOR_ASSIGNMENT);
+
+  private static final Set<String> FUNCTION_CHANGING_CURRENT_SCOPE = new HashSet<>(Arrays.asList(
+    "eval",
+    "extract",
+    "parse_str",
+    // PREG_REPLACE_EVAL option was deprecated in php 5.5 and has been removed in php 7.0
+    "preg_replace",
+    "include",
+    "include_once",
+    "require",
+    "require_once"));
+
+  // Note: "$argc" and "$argv" are not available in the function scope without using "global"
+  private static final Set<String> PREDEFINED_VARIABLES = new HashSet<>(Arrays.asList(
+    "$_cookie",
```

Comment:
According to my tests, it seems that predefined variable names are case-sensitive.  Here it should be `$_COOKIE`.